### DISCUSSION TITLE
What is the state of deno compatibility?

### DISCUSSION THREAD

**Discussion by jpambrun** [denoland/deno#23265] [Q&A]
I am curious about deno and I am trying to convert a simple script, but I am hitting a snag at the very first line.

``` js
import * as dicomStream from "npm:@exini/dicom-streams-js";
```
Which gives:
```
error: Could not find npm package 'zlib' referenced by '@exini/dicom-streams-js@3.0.9'.
```

on
```sh
❯ deno --version
deno 1.42.1 (release, x86_64-unknown-linux-gnu)
v8 12.3.219.9
typescript 5.4.3
```

Perousing the code I had the feeling deno should provide the necessary polyfill for zlib? Am I missing something? Aren't node modules loaded with `npm:` supposed to have these shims enabled?

**Comment by bartlomieju** [comment]
This looks like a resolution bug. Deno supports built-in Node, `zlib` module.

I need to debug more, but the way it's "require"d is a bit funky:
```js
...
Object.defineProperty(exports, "__esModule", ({ value: true }));
exports.deflateDatasetFlow = exports.toUtf8Flow = exports.toIndeterminateLengthSequences = exports.fmiGroupLengthFlow = exports.validateContextFlow = exports.ValidationContext = exports.headerFilter = exports.fmiDiscardFilter = exports.groupLengthDiscardFilter = exports.denyFilter = exports.allowFilter = exports.tagFilter = exports.stopTagFlow = exports.toBytesFlow = void 0;
const stream_1 = __webpack_require__(/*! stream */ "stream");
const zlib_1 = __importDefault(__webpack_require__(/*! zlib */ "zlib"));
const base_1 = __webpack_require__(/*! ./base */ "./src/base.ts");
const character_sets_1 = __webpack_require__(/*! ./character-sets */ "./src/character-sets.ts");
...
```

It might have to do with `__importDefault()` - "streams" is only "require"d an works fine.

**Reply by jpambrun** [reply]
I see, that does look very challenging. I also just notice some poeple started publishing bun specific package on npm.. which will make it a even more complicated and confusing, eg. https://www.npmjs.com/package/bun-compression.

**Reply by bartlomieju** [reply]
Okay, nice one! I found the problem here. The problematic part is this:
```js
/***/ "uuid":
/*!***********************!*\
  !*** external "uuid" ***!
  \***********************/
/***/ ((module) => {

module.exports = require("uuid");

/***/ }),

/***/ "zlib":
/*!***********************!*\
  !*** external "zlib" ***!
  \***********************/
/***/ ((module) => {

    module.exports = require("zlib");

    /***/ }),

/***/ "stream":
/*!*************************!*\
  !*** external "stream" ***!
  \*************************/
/***/ ((module) => {

module.exports = require("stream");

/***/ }),
```
When we analyze CommonJS file to translate them to ESM we only discover the very last reexport here (which also seems wrong, because it's not a reexport of the top level module, since it's a bundled file).

I opened https://github.com/denoland/deno_ast/issues/227 to fix it.

After I patched it in our internal code to skip this particular module, the module imports just fine 👍

**Comment by ffMathy** [comment]
I made this site to track Deno compatibility every day: https://github.com/ffMathy/is-deno-compatible-yet
